src/Sylius/Bundle/ShopBundle/Grid/ProductGrid.php

```diff
@@ -0,0 +1,111 @@
+<?php
+
+/*
+ * This file is part of the Sylius package.
+ *
+ * (c) Sylius Sp. z o.o.
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+declare(strict_types=1);
+
+namespace Sylius\Bundle\ShopBundle\Grid;
+
+use Sylius\Bundle\GridBundle\Builder\Field\DateTimeField;
+use Sylius\Bundle\GridBundle\Builder\Field\Field;
+use Sylius\Bundle\GridBundle\Builder\Field\StringField;
+use Sylius\Bundle\GridBundle\Builder\Filter\Filter;
+use Sylius\Bundle\GridBundle\Builder\GridBuilderInterface;
+use Sylius\Bundle\GridBundle\Grid\AbstractGrid;
+use Sylius\Bundle\GridBundle\Grid\ResourceAwareGridInterface;
+use Sylius\Component\Channel\Context\ChannelContextInterface;
+use Sylius\Component\Core\Model\TaxonInterface;
+use Sylius\Component\Locale\Context\LocaleContextInterface;
+use Sylius\Component\Taxonomy\Repository\TaxonRepositoryInterface;
+use Symfony\Component\HttpFoundation\RequestStack;
+use Symfony\Component\HttpKernel\Exception\NotFoundHttpException;
+use Webmozart\Assert\Assert;
+
+final class ProductGrid extends AbstractGrid implements ResourceAwareGridInterface
+{
+    /**
+     * @param TaxonRepositoryInterface<TaxonInterface> $taxonRepository
+     */
+    public function __construct(
+        private string $resourceClass,
+        private ChannelContextInterface $channelContext,
+        private LocaleContextInterface $localeContext,
+        private TaxonRepositoryInterface $taxonRepository,
+        private RequestStack $requestStack,
+        private bool $includeAllDescendants,
+    ) {
+    }
+
+    public static function getName(): string
+    {
+        return 'sylius_shop_product';
+    }
+
+    public function buildGrid(GridBuilderInterface $gridBuilder): void
+    {
+        $request = $this->requestStack->getMainRequest();
+        Assert::notNull($request, 'No main request available.');
+
+        $localeCode = $this->localeContext->getLocaleCode();
+
+        $slug = $request->attributes->get('slug', '');
+        if (!is_string($slug)) {
+            throw new NotFoundHttpException('The request attribute "slug" must be a string.');
+        }
+
+        // @see Sylius\Bundle\ResourceBundle\ExpressionLanguage\NotNullExpressionFunctionProvider
```

Comment:
IMO `@see` message is not needed, check for taxon presence is understandable

Reply:
Yeah, I thought because it was the same code I copied over I left a reference but that is better.